### PR TITLE
Make serialization work with backing fields automatically

### PR DESCRIPTION
**Pull request by DrSmugleaf**
This removes the need to write [field: DataField(...)] on get-only properties with a backing field, for which the only warning was a line in the serialization doc and an error after starting up the game.

The backing field is found by looking for a field with the name \<PropertyName\>k__BackingField and an attribute of CompilerGenerated, as far as I know these are an implementation detail but the same across C# and a bunch of other compilers. The method that finds the backing field is in TypeHelpers in case its implementation ever needs to be changed. The other way to do it would be to analyze the instructions of the method body and finding the Ldfld op code (second instruction) and going from there, but that is above my paygrade.

### CODE REVIEW

File: Robust.Shared/Serialization/Manager/SerializationDataDefinition.cs [hunk 1]
```diff
@@ -87,7 +88,12 @@ public SerializationDataDefinition(Type type)
             {
                 var attr = abstractFieldInfo.GetCustomAttribute<DataFieldAttribute>();
 
-                if (attr == null) continue;
+                if (attr == null || abstractFieldInfo.IsBackingField())
```
Comment (PaulRitter) [inline Robust.Shared/Serialization/Manager/SerializationDataDefinition.cs:90]: this disables the [field: ]-variant. why?

File: Robust.Shared/Utility/TypeHelpers.cs [hunk 2]
```diff
@@ -281,10 +282,56 @@ public static IEnumerable<AbstractFieldInfo> GetAllPropertiesAndFields(this Type
 
             return commonType;
         }
+
+        public static SpecificFieldInfo? GetBackingField(this Type type, string propertyName)
+        {
+            foreach (var parent in type.GetClassHierarchy())
+            {
+                var field = parent.GetField($"<{propertyName}>k__BackingField",
+                    BindingFlags.NonPublic | BindingFlags.Instance);
```
Comment (PaulRitter) [inline Robust.Shared/Utility/TypeHelpers.cs:291]: i am 99% sure there is a property on the propertyinfo object for this

File: Robust.Shared/Serialization/Manager/SerializationDataDefinition.cs [hunk 3]
```diff
@@ -140,7 +160,27 @@ public SerializationDataDefinition(Type type)
...
+                if (!Equals(definition.Attribute, attribute))
+                {
+                    continue;
+                }
+
+                return definition.InheritanceBehaviour;
+            }
+
+            return null;
+        }
```
Comment (PaulRitter) [inline Robust.Shared/Serialization/Manager/SerializationDataDefinition.cs]: please yeet this out of the class and into the type, make it an extension or smth. exposing basefielddefinitions as a readonly list is cleaner than having this in the class imo

**Review by PaulRitter — CHANGES_REQUESTED** [review]
i only have limited braincells due to exams but looks ok aside from the few notes i had.
i also saw you using generics and nullability and got a bit scared but as i said my mental capacity is nearing a 5 year low point so i just skipped that part and trusted you. pls dont disappoint

File: Robust.Shared/Serialization/Manager/SerializationDataDefinition.cs (hunk 1 above)
Comment (DrSmugleaf) [reply Robust.Shared/Serialization/Manager/SerializationDataDefinition.cs:90]: That is to skip registering both the property and the backing field, though you are right if field: is used then it should not be skipped.

File: Robust.Shared/Serialization/Manager/SerializationDataDefinition.cs (hunk 3 above)
Comment (DrSmugleaf) [reply Robust.Shared/Serialization/Manager/SerializationDataDefinition.cs]: It is only for the test but will do.

File: Robust.Shared/Utility/TypeHelpers.cs (hunk 2 above)
Comment (PaulRitter) [reply Robust.Shared/Utility/TypeHelpers.cs:291]: i was 99% wrong disregard

File: Robust.Shared/Serialization/Manager/SerializationDataDefinition.cs (hunk 1 above)
Comment (DrSmugleaf) [reply Robust.Shared/Serialization/Manager/SerializationDataDefinition.cs:90]: On second thought this is already a test case. I changed it so looking up attributes on the property object also returns the ones on its backing field.

File: Robust.Shared/Serialization/Manager/SerializationDataDefinition.cs (hunk 3 above)
Comment (DrSmugleaf) [reply Robust.Shared/Serialization/Manager/SerializationDataDefinition.cs]: Done in https://github.com/space-wizards/RobustToolbox/pull/1727/commits/5aa025da6611792b0a83a7190f27c670f3a30538

File: Robust.UnitTesting/Shared/Serialization/PropertyAndFieldDefinitionTest.cs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            mapping.AddNode(GetAndSetPropertyName, new ValueDataNode("15"));
+            mapping.AddNode(FieldName, new ValueDataNode("20"));
+            mapping.AddNode(GetOnlyPropertyWithOtherAttributeFieldTargetedName, new ValueDataNode("25"));
+            mapping.AddNode(GetOnlyPropertyFieldTargetedAndOtherAttributeName, new ValueDataNode("30"));
+
+            var definition = Serialization.ReadValue<PropertyAndFieldDefinitionTestDefinition>(mapping);
+
+            Assert.NotNull(definition);
+
+            // Set only property with backing field, property targeted
```
Comment (RemieRichards) [inline Robust.UnitTesting/Shared/Serialization/PropertyAndFieldDefinitionTest.cs:41]: This comment, and the one below, refer to a "Set only" property.
Yet the two asserts are against a get only property.

So 1 of 3 things it at play.

1. The comments are wrong (i.e. "Set" vs "Get")
2. The code is wrong (Using wrong field, field defined wrong, etc.)
3. The comments are misleading and should be reworded (Halp brain melt)

Comment (DrSmugleaf) [reply Robust.UnitTesting/Shared/Serialization/PropertyAndFieldDefinitionTest.cs:41]: 1 was correct, the comments were wrong. I have fixed up some of the other comments as well to make them consistent with the rest.
Fixed in https://github.com/space-wizards/RobustToolbox/commit/61113d24349c6e2a10b4b908e17c3e87d0a9a34c

**Pull request merged**